### ISSUE TITLE
Add a "page rank" to guides

### ISSUE THREAD

**Issue by yrodiere** [quarkusio/search.quarkus.io#12]
See https://github.com/quarkusio/search.quarkus.io/pull/11/files/be54fbcb07484887a179c56e9bf31d1177b72e00#diff-cf87147ed8cc692330d20b5a18f42b8a3957a65f0108d0da117344f17720dcfd

Basically we want the Hibernate ORM guide to be boosted so that it appears before e.g. the Hibernate Search guide, because it's more mainstream.

We could achieve this with a rank feature field in Elasticsearch, computing some kind of "page rank" by determining which guide refers another guide.

**Comment by yrodiere** [comment]
See https://en.wikipedia.org/wiki/PageRank#Relevant_patents , though a simplified version will probably suffice.

Note in particular:

> As of September 24, 2019, all patents associated with PageRank have expired.

**Comment by yrodiere** [comment]
Here is pseudo code for some pagerank implementation that will probably be more than enough for our use case: https://spark.apache.org/docs/preview/api/java/org/apache/spark/graphx/lib/PageRank.html

**Comment by yrodiere** [comment]
IMPORTANT: We could probably achieve something "good enough" simply by hardcoding the ranks in guide metadata. We don't necessarily need to implement pagerank.

**Comment by yrodiere** [comment]
> IMPORTANT: We could probably achieve something "good enough" simply by hardcoding the ranks in guide metadata. We don't necessarily need to implement pagerank.

Hey @gsmet would that make sense?

**Comment by gsmet** [comment]
Yeah, I think we should have some sort or rank or priority. I actually wanted to implement the priority idea to improve the index page and put the most relevant guides first.

Now what I'm unsure is that for search, wouldn't you need a rank per topic? Or you think it will be good enough?

I.e. won't you have guides prioritized for a search that is not really relevant to it but still matches?

**Comment by gsmet** [comment]
> I'm not sure what you have in mind for "a rank per topic", though... perhaps you have a better idea.

Well, no, I don't have a better idea. I'm just wondering if it will be good enough but... I think we should implement it and see, rather than waiting for a better idea we probably won't have :)

**Comment by yrodiere** [comment]
> The impact of ranks, be they manual or pagerank, will definitely be independent from the "natural" score of the query (based on tokens). So there will be cases where we boost a first-class guide and end up giving it a better position in results than another guide that was precisely what the user wanted.

Actually the impact can be multiplicative instead of additive if we use this: https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-function-score-query.html
Which will be a pain to implement (Hibernate Search doesn't support wrapping "built-in" predicates with custom json), but could do what we want... Assuming it doesn't _completely_ ruin performance (it disables some optimizations related to non-competitive hit skipping :/ )
